solidity/contracts/liquidity-protection/TokenTimeWeightedAverage.sol

```diff
@@ -0,0 +1,375 @@
+        require(hasRole(ROLE_SEEDER, msg.sender), "ERR_ACCESS_DENIED");
+
+        uint256 length = _ns.length;
+        require(length == _ds.length && length == _times.length, "ERR_INVALID_LENGTH");
+
+        for (uint256 i = 0; i < length; ++i) {
+            uint256 n = _ns[i];
+            uint256 d = _ds[i];
+            uint256 t = _times[i];
+
+            _greaterThanZero(d);
+            require(t < time(), "ERR_INVALID_TIME");
+
+            addSample(_token, Fraction({ n: n, d: d }), t);
+        }
+    }
+
+    /**
+     * @dev calculated the TWA from a specific sample time
+     *
+     * @param _token the token the data is accumulated for
+     * @param _startTime the sampling starting time
+     *
+     * @return TWA's numerator and denominator
+     */
+    function timeWeightedAverage(IERC20Token _token, uint256 _startTime)
+        external
+        view
+        override
+        validAddress(address(_token))
+        returns (uint256, uint256)
+    {
+        TokenData storage tokenData = data[_token];
+        require(tokenData.timestamps[_startTime], "ERR_NO_DATA");
+
+        return timeWeightedAverage(tokenData, _startTime, tokenData.lastSampleTime);
+    }
+
+    /**
+     * @dev calculated the TWA between to specific sample times
+     *
+     * @param _token the token the data is accumulated for
+     * @param _startTime the sampling starting time
+     * @param _endTime the sampling ending time
+     *
+     * @return TWA's numerator and denominator
+     */
+    function timeWeightedAverage(
+        IERC20Token _token,
+        uint256 _startTime,
+        uint256 _endTime
+    ) external view override validAddress(address(_token)) returns (uint256, uint256) {
+        require(_startTime < _endTime, "ERR_INVALID_TIME");
+
+        TokenData storage tokenData = data[_token];
+        require(tokenData.timestamps[_startTime] && tokenData.timestamps[_endTime], "ERR_NO_DATA");
+
+        return timeWeightedAverage(tokenData, _startTime, _endTime);
+    }
+
+    /**
+     * @dev calculated the TWA between to specific sample times
+     *
+     * @param _tokenData the token accumulation data
+     * @param _startTime the sampling starting time
+     * @param _endTime the sampling ending time
+     *
+     * @return TWA's numerator and denominator
+     */
+    function timeWeightedAverage(
+        TokenData storage _tokenData,
+        uint256 _startTime,
+        uint256 _endTime
+    ) private view returns (uint256, uint256) {
+        Fraction memory endAccumulator = _tokenData.accumulators[_endTime];
+
+        // if we have received only a single sample - just return it
+        if (_tokenData.firstSampleTime == _endTime) {
+            return (endAccumulator.n, endAccumulator.d);
+        }
+
+        Fraction memory startAccumulator = _tokenData.accumulators[_startTime];
+
+        // TWA = (endAccumulator - startAccumulator) / (_endTime - _startTime)
```

Comment:
Better yet:
```
uint256 n = startAccumulator.d.mul(endAccumulator.n).sub(startAccumulator.n.mul(endAccumulator.d));
uint256 d = startAccumulator.d.mul(endAccumulator.d).mul(_endTime.sub(_startTime));
```
No need to lose precision on division here.

Reply:
👍

Reply:
As you can understand, I only moved the "div" part from the numerator to be a "mul" part in the denominator.
You may notice a few other changes in each one of them, but those are only cosmetic (OCD-related) reorders of the multiplication operands.